examples/efficientdet/efficientdet_building_blocks.py

```diff
@@ -0,0 +1,321 @@
+import functools
+
+import tensorflow as tf
+from config import get_fpn_configuration
+
+
+class ResampleFeatureMap(tf.keras.layers.Layer):
+    """Resample feature maps for downsampling or upsampling
+     to create coarser or finer additional feature maps."""
+
+    def __init__(self,
+                 feat_level,
+                 target_num_channels,
+                 apply_bn,
+                 conv_after_downsample,
+                 data_format='channels_last',
+                 pooling_type=None,
+                 upsampling_type=None,
+                 name='resample_p0'):
+
+        super().__init__(name=name)
+        self.apply_bn = apply_bn
+        self.data_format = data_format
+        self.target_num_channels = target_num_channels
+        self.feat_level = feat_level
+        self.conv_after_downsample = conv_after_downsample
+        self.pooling_type = pooling_type or 'max'
+        self.upsampling_type = upsampling_type or 'nearest'
+        self.bn = tf.keras.layers.BatchNormalization(momentum=0.99,
+                                                     epsilon=1e-3)
+        self.conv2d = tf.keras.layers.Conv2D(self.target_num_channels,
+                                             (1, 1),
+                                             padding='same',
+                                             data_format=self.data_format,
+                                             name='conv2d')
+
+    def _pool2d(self, inputs, height, width, target_height, target_width):
+        """Pool the inputs to target height and width."""
+        height_stride_size = int((height - 1) // target_height + 1)
+        width_stride_size = int((width - 1) // target_width + 1)
+        if self.pooling_type == 'max':
+            return tf.keras.layers.MaxPooling2D(
+                pool_size=[height_stride_size + 1, width_stride_size + 1],
+                strides=[height_stride_size, width_stride_size],
+                padding='SAME',
+                data_format=self.data_format)(inputs)
+        if self.pooling_type == 'avg':
+            return tf.keras.layers.AveragePooling2D(
+                pool_size=[height_stride_size + 1, width_stride_size + 1],
+                strides=[height_stride_size, width_stride_size],
+                padding='SAME',
+                data_format=self.data_format)(inputs)
+        raise ValueError('Unsupported pooling type {}.'.format(
+            self.pooling_type))
+
+    def _upsample2d(self, inputs, target_height, target_width):
+        return tf.cast(
+            tf.compat.v1.image.resize_nearest_neighbor(
```

Comment:
In Keras, nearest and bilinear both are available in the upsample2d API. This is the resize used in the official implementation of EfficientDet and need not be incorporated in the graph because this resize is used to match only the feature map sizes from different feature levels. Once, the size mismatch is addressed they can be fused. So, differentiability is not required here.